zerver/lib/export.py

```diff
@@ -672,32 +672,86 @@ def get_ids(records):
         # type: (List[Record]) -> Set[int]
         return set(x['id'] for x in records)
 
-    # TODO: We need to properly filter messages that were only
-    #       sent among mirrordummy users here.
-    user_profile_ids = get_ids(response['zerver_userprofile'] +
-                               response['zerver_userprofile_mirrordummy'] +
-                               response['zerver_userprofile_crossrealm'])
-    recipient_ids = get_ids(response['zerver_recipient'])
-
-    # Basic security rule: You can export everything sent by someone
-    # in your realm export (members of your realm plus Zulip realm
-    # bots) to a recipient object you're exporting (that is thus also
-    # in your realm).
+    # Basic security rule: You can export everything either...
+    #   - sent by someone in your exportable_user_ids
+    #        OR
+    #   - received by someone in your exportable_user_ids (which
+    #     equates to a recipient object we are exporting)
     #
     # TODO: In theory, you should be able to export messages in
     # cross-realm PM threads; currently, this only exports cross-realm
     # messages received by your realm that were sent by Zulip system
     # bots (e.g. emailgateway, notification-bot).
-    message_query = Message.objects.filter(sender__in=user_profile_ids,
-                                           recipient__in=recipient_ids).order_by("id")
+
+    realm_subscriptions = Subscription.objects.filter(user_profile__realm=realm)
+    realm_subscription_ids = set([row.id for row in realm_subscriptions])
+    realm_recipients = Recipient.objects.filter(subscription__in=realm_subscription_ids)
+
+    # Here, "we" and "us" refers to the inner circle of users who
+    # were specified as being allowed to be exported.  "Them"
+    # refers to other users.
+    user_ids_for_us = get_ids(
+        response['zerver_userprofile']
+    )
+    recipient_ids_for_us = get_ids(response['zerver_recipient'])
+
+    ids_of_our_possible_senders = get_ids(
+        response['zerver_userprofile'] +
+        response['zerver_userprofile_mirrordummy'] +
+        response['zerver_userprofile_crossrealm'])
+
+    all_realm_recipient_ids = set([r.id for r in realm_recipients])
+    recipient_ids_for_them = all_realm_recipient_ids - recipient_ids_for_us
```

Comment:
If I'm not mistaken, I think that `recipient_ids_for_them` can be computed with just:

`recipient.objects.filter(type=Recipient.PERSONAL, type_id__in=ids_of_our_possible_senders).values("id")`

Since as we've discussed, messages sent by anyone in `user_ids_for_us` to a huddle or stream will already be in `messages_we_received` (because `user_ids_for_us` is a subset of `ids_of_our_possible_senders`, and you can only send to stream/huddles you are a member of).